akka-bench-jmh/src/main/scala/akka/http/HttpResponseRenderingBenchmark.scala

```diff
@@ -0,0 +1,249 @@
+
+  val requestRendering: Flow[HttpRequest, String, NotUsed] =
+    Http()
+      .clientLayer(headers.Host("blah.com"))
+      .atop(TlsPlacebo)
+      .join {
+        Flow[ByteString].map { x ⇒
+          val response = s"HTTP/1.1 200 OK\r\nContent-Length: ${x.size}\r\n\r\n"
+          ByteString(response) ++ x
+        }
+      }
+      .mapAsync(1)(response => Unmarshal(response).to[String])
+
+  def renderResponse: Future[String] = Source.single(HttpRequest(uri = "/foo"))
+    .via(requestRendering)
+    .runWith(Sink.head)
+
+  var request: HttpRequest = _
+  var pool: Flow[(HttpRequest, Int), (Try[HttpResponse], Int), _] = _
+
+  @TearDown
+  def shutdown(): Unit = {
+    Await.ready(Http().shutdownAllConnectionPools(), 1.second)
+    Await.result(system.terminate(), 5.seconds)
+  }
+
+  /*
+  [info] Benchmark                                               Mode  Cnt              Score              Error  Units
+  [info] HttpResponseRenderingBenchmark.header_date_val         thrpt   20  2 704 169 260 029.906 ± 234456086114.237  ops/s
+  
+  // def, normal time
+  [info] HttpResponseRenderingBenchmark.header_date_def             thrpt   20    178 297 625 609.638 ± 7429280865.659  ops/s
+  [info] HttpResponseRenderingBenchmark.response_ok_simple_val      thrpt   20          1 258 119.673 ± 58399.454  ops/s
+  [info] HttpResponseRenderingBenchmark.response_ok_simple_def      thrpt   20            687 576.928 ± 94813.618  ops/s
+  
+  // clock nanos
+  [info] HttpResponseRenderingBenchmark.response_ok_simple_clock    thrpt   20          1 676 438.649 ± 33976.590  ops/s
+  [info] HttpResponseRenderingBenchmark.response_ok_simple_clock    thrpt   40          1 199 462.263 ± 222226.304  ops/s
+  
+  // ------
+  
+  // before optimisig collectFirst
+  [info] HttpResponseRenderingBenchmark.json_response        thrpt   20  1 782 572.845 ± 16572.625  ops/s
+  [info] HttpResponseRenderingBenchmark.simple_response      thrpt   20  1 611 802.216 ± 19557.151  ops/s
+  
+  // after removing collectFirst and Option from renderHeaders
+  // not much of a difference, but hey, less Option allocs
+  [info] HttpResponseRenderingBenchmark.json_response        thrpt   20  1 785 152.896 ± 15210.299  ops/s
+  [info] HttpResponseRenderingBenchmark.simple_response      thrpt   20  1 783 800.184 ± 14938.415  ops/s
+  
+  // -----
+
+  // baseline for this optimisation is the above results (after collectFirst).
+  
+  // after introducing pre-rendered ContentType headers:
+  
+  normal clock
+  [info] HttpResponseRenderingBenchmark.json_long_raw_response  thrpt   20  1738558.895 ± 159612.661  ops/s
+  [info] HttpResponseRenderingBenchmark.json_response           thrpt   20  1714176.824 ± 100011.642  ops/s
+
+  "fast clock"
+  [info] HttpResponseRenderingBenchmark.json_long_raw_response  thrpt   20  1 528 632.480 ± 44934.827  ops/s
+  [info] HttpResponseRenderingBenchmark.json_response           thrpt   20  1 517 383.792 ± 28256.716  ops/s
```

Comment:
I realised the clock benchmarking wasn't realistic since the gain is from
accessing it from multiple threads but we never do that really. I'll
optimise it to be a shared thing!

-- 
Konrad `ktoso` Malawski
Akka <http://akka.io> @ Lightbend <http://lightbend.com>

On 28 July 2016 at 15:04:32, Johan Andrén (notifications@github.com) wrote:

In
akka-bench-jmh/src/main/scala/akka/http/HttpResponseRenderingBenchmark.scala
<https://github.com/akka/akka/pull/21046#discussion_r72617384>:

> +  [info] HttpResponseRenderingBenchmark.json_response        thrpt   20  1 785 152.896 ± 15210.299  ops/s
> +  [info] HttpResponseRenderingBenchmark.simple_response      thrpt   20  1 783 800.184 ± 14938.415  ops/s
> +
> +  // -----
> +
> +  // baseline for this optimisation is the above results (after collectFirst).
> +
> +  // after introducing pre-rendered ContentType headers:
> +
> +  normal clock
> +  [info] HttpResponseRenderingBenchmark.json_long_raw_response  thrpt   20  1738558.895 ± 159612.661  ops/s
> +  [info] HttpResponseRenderingBenchmark.json_response           thrpt   20  1714176.824 ± 100011.642  ops/s
> +
> +  "fast clock"
> +  [info] HttpResponseRenderingBenchmark.json_long_raw_response  thrpt   20  1 528 632.480 ± 44934.827  ops/s
> +  [info] HttpResponseRenderingBenchmark.json_response           thrpt   20  1 517 383.792 ± 28256.716  ops/s

so fast clock is a bit worse?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/akka/akka/pull/21046/files/6d85986c2e45eaaf7e43a0249810e0449c86e983#r72617384>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/AAHYkyDcIzsNNJpbZaO-WcWlAU42WGMCks5qaKjggaJpZM4JViR1>
.